CHERI-RISC-V Mode-Dependent Jumps + Compressed Instructions

jrtc27 opened the pull request:
Things to note:

1. I haven't added any encoding tests for the compressed instructions (nor predicate enforcement tests) as they're tedious and I never got round to it; personally I think it'd be better to land this so we can get on with evaluating, but if others feel strongly that can become a blocker.
2. ~~Old-style function call codegen is no longer supported and so hardware with mode-dependent jumps is required, thus this should wait until that's landed in sufficiently many implementations (and the others will then be forced to catch up, though it's not a particularly big change to make).~~ There's now a cl::opt to use the old indirect function call sequence, though it'll still use the new mode-dependent CJALR rather than JALR.CAP to do the actual jump.
3. Compressed instructions are deliberately off by default (gated by -mxcheri-rvc), as those are more work to support, especially in Flute.
4. No lazy binding as it's a pain while we don't use the real LLD GOT.
5. ~I haven't added an R_RISCV_CHERI_RVC_JUMP yet, so RV32 might be broken if you end up with a compressed JAL to a symbol, though such assembly is buggy and should be using (C)TAIL (and you have to have opted in to -mxcheri-rvc). Easy enough to fix though, just forgot about it until I revisited this branch.~ Fixed.

I do however think it's good enough to land, bar review comments, once hardware is ready, but perhaps I should fix 5 in the meantime for completeness and cleanliness.

arichardson commented on lld/test/ELF/cheri/riscv/plt.s:
```diff
@@ -0,0 +1,94 @@
+
+# DIS:      <_start>:
+## Direct call
+## foo - . = 0x11020-0x11000 = 32
```
I think it would be good to also check the section addresses for the ones used in these calculations. I can see that one CHECK line verifies that .captable is at 0x12000, but it would also be good to have CHECK lines for the constants used in calculations here.

jrtc27 replied:
We do from the disassembly; all the referenced constants have DIS lines that match them. This one for example has:
```
# DIS:      <foo>:
# DIS-NEXT:   11020:
```
below to check the LHS, and 
```
# DIS-NEXT: 11000: auipcc cra, 0
```
on the line just below this one to check the RHS.

arichardson commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVAsmBackend.cpp:
```diff
@@ -175,24 +181,38 @@ void RISCVAsmBackend::relaxInstruction(MCInst &Inst,
-    Res.setOpcode(RISCV::JAL);
-    Res.addOperand(MCOperand::createReg(RISCV::X0));
-    Res.addOperand(Inst.getOperand(0));
+    // c.j $imm -> (c)jal X0, $imm.
+    if (IsCapMode) {
+      Res.setOpcode(RISCV::CJAL);
+      Res.addOperand(MCOperand::createReg(RISCV::C0));
+      Res.addOperand(Inst.getOperand(0));
+    } else {
+      Res.setOpcode(RISCV::JAL);
+      Res.addOperand(MCOperand::createReg(RISCV::X0));
+      Res.addOperand(Inst.getOperand(0));
+    }
```
```suggestion
    // c.j $imm -> (c)jal X0, $imm.
    Res.setOpcode(IsCapMode ? RISCV::CJAL : RISCV::JAL);
    Res.addOperand(
      MCOperand::createReg(IsCapMode ? RISCV::C0 : RISCV::X0));
    Res.addOperand(Inst.getOperand(0));
```
Maybe like this to make it a bit shorter?

jrtc27 replied:
It's shorter but I find it harder to read and easier to mess up like that given the condition is duplicated, at least for me

jrtc27 replied:
Though I guess I'm inconsistent as I did that for the expansion of (C)CALL

arichardson reviewed (approved):
I haven't check the encodings of the newly added instructions, but otherwise this looks good 👍

jrtc27 commented:
Rebased and added a -riscv-legacy-indirect-purecap-calls option that restores the old always-make-an-indirect-call behaviour. It will still lower the actual indirect call using the new mode-dependent CJALR-with-immediate instruction, though, just like any other indirect call will. So long as the hardware supports the instruction, that should make no difference (with the exception of compression, but that can be separately configured with -mxcheri-rvc).

arichardson commented:
Is this ready to be merged or are we still waiting for hardware support?

PeterRugg commented:
> Is this ready to be merged or are we still waiting for hardware support?

All the hardware supports this. The only potential holdup is that I've been meaning to do a systematic review of the encodings but still haven't gotten around to it. Since everything seems to work, it should be okay but I'll try to do it now and hit approve.

PeterRugg commented on llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:
```diff
@@ -838,7 +890,201 @@ def : InstAlias<"cfsd $rs2, (${rs1})", (CFSD FPR64:$rs2, GPCR:$rs1, 0), 0>;
+let DecoderNamespace = "RISCV32CapModeOnly_",
+    Predicates = [HasCheri, HasCheriRVC, HasStdExtC, HasStdExtD, IsRV32, IsCapMode] in
+def C_CFLDCSP : CCheriStackLoad<0b001, "c.cfldcsp", FPR64, uimm9_lsb000> {
+  let Inst{6-5} = imm{4-3};
```
let Inst{12} = imm{5} ?

jrtc27 replied:
Yep, seems I have indeed missed that

jrtc27 replied:
Oh, no, I see why. Because every CI-format instruction has bit 5 of the immediate in bit 12 of the instruction, that's factored out into RVInst16CI's definition. And I'm pretty sure TableGen shouts at you when you've left bits of the encoding uninitialised, but can't say for certain.

jrtc27 replied:
Interestingly stack stores (CSS-format) don't do the same thing, I guess because there's always at least one other bit that's contiguous so it doesn't save any lines of code as bit 12/5 can always be folded into an existing line in the instruction definition, but still a bit inconsistent.

PeterRugg replied:
Okay, sounds good

PeterRugg commented on llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:
```diff
@@ -838,7 +890,201 @@ def : InstAlias<"cfsd $rs2, (${rs1})", (CFSD FPR64:$rs2, GPCR:$rs1, 0), 0>;
+let Predicates = [HasCheri, HasCheriRVC, HasStdExtC, IsRV64, IsCapMode] in
+def C_CLCCSP_128 : CCheriStackLoad<0b001, "c.clccsp", GPCRNoC0, uimm10_lsb0000> {
+  let Inst{6} = imm{4};
+  let Inst{5-2} = imm{9-6};
```
let Inst{12} = imm{5} ?

PeterRugg commented on llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:
```diff
@@ -838,7 +890,201 @@ def : InstAlias<"cfsd $rs2, (${rs1})", (CFSD FPR64:$rs2, GPCR:$rs1, 0), 0>;
+}
+
+def C_CLWCSP : CCheriStackLoad<0b010, "c.clwcsp", GPRNoX0, uimm8_lsb00> {
+  let Inst{6-4} = imm{4-2};
```
let Inst{12} = imm{5} ?

PeterRugg commented on llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:
```diff
@@ -838,7 +890,201 @@ def : InstAlias<"cfsd $rs2, (${rs1})", (CFSD FPR64:$rs2, GPCR:$rs1, 0), 0>;
+let DecoderNamespace = "RISCV32CapModeOnly_",
+    Predicates = [HasCheri, HasCheriRVC, HasStdExtC, IsRV32, IsCapMode] in
+def C_CLCCSP_64 : CCheriStackLoad<0b011, "c.clccsp", GPCRNoC0, uimm9_lsb000> {
+  let Inst{6-5} = imm{4-3};
```
let Inst{12} = imm{5} ?

jrtc27 commented on llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:
```diff
@@ -601,11 +614,46 @@ defm AMOSWAP_C  : AMO_C_rr_aq_rl<"128", 0b00001, 0b100, "amoswap.c", GPCR>;
+
+let Predicates = [HasCheri, IsCapMode] in {
+def : InstAlias<"j $offset",    (CJAL C0, simm21_lsb0_jal:$offset)>;
+def : InstAlias<"cjal $offset", (CJAL C1, simm21_lsb0_jal:$offset)>;
```
Yeah, it has an implicit capability destination so the consistent thing to do is use cjal.

PeterRugg commented on llvm/lib/Target/RISCV/RISCVInstrInfoXCheri.td:
```diff
@@ -838,7 +890,201 @@ def : InstAlias<"cfsd $rs2, (${rs1})", (CFSD FPR64:$rs2, GPCR:$rs1, 0), 0>;
+
+let Predicates = [HasCheri, HasCheriRVC, HasStdExtC, IsRV64, IsCapMode] in
+def C_CLDCSP : CCheriStackLoad<0b011, "c.cldcsp", GPRNoX0, uimm9_lsb000> {
+  let Inst{6-5} = imm{4-3};
```
let Inst{12} = imm{5} ?

PeterRugg commented:
I've checked everything I can see, and apart from immediate bit 5 in the cases I've highlighted, looks good to me :)

jrtc27 commented on llvm/lib/Target/RISCV/AsmParser/RISCVAsmParser.cpp:
```diff
@@ -445,6 +447,16 @@ struct RISCVOperand : public MCParsedAsmOperand {
   }
 
+  bool isCCallSymbol() const {
+    int64_t Imm;
+    RISCVMCExpr::VariantKind VK = RISCVMCExpr::VK_RISCV_None;
+    // Must be of 'immediate' type but not a constant.
+    if (!isImm() || evaluateConstantImm(getImm(), Imm, VK))
+      return false;
+    return RISCVAsmParser::classifySymbolRef(getImm(), VK) &&
+           VK == RISCVMCExpr::VK_RISCV_CCALL;
```
I thought about it but upstream has so many copies of the thing that I don't want to change them and get conflicts all the time, so figured it was best to just go with the flow

jrtc27 commented on llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp:
```diff
@@ -116,39 +116,68 @@ void RISCVMCCodeEmitter::expandFunctionCall(const MCInst &MI, raw_ostream &OS,
   MCInst TmpInst;
   MCOperand Func;
   MCRegister Ra;
+  bool IsCap;
```
Same is true for Ra, and either true for Func or it has a default constructor so the assert would fire. I can see the argument for the default init wrt future merges, but it would look a bit out of place...

jrtc27 replied:
New entries here should presumably give merge conflicts though so won't silently use uninitialised data?

arichardson reviewed (approved):
Let's wait for Jenkins, but this looks good!
